File: backend/app/routers/admin/properties.py
from datetime import date, timedelta

from fastapi import APIRouter, Depends, HTTPException, Response
from sqlalchemy import delete, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.security import AdminClaims, require_super_admin
from app.db import get_session
from app.models import Availability, Booking, Property, RatePlan, RoomType
from app.schemas.admin_content import (
    AvailabilityBulkUpsert,
    PropertyCreate,
    PropertyUpdate,
    RatePlanCreate,
    RatePlanUpdate,
    RoomTypeCreate,
    RoomTypeUpdate,
)

MAX_AVAILABILITY_RANGE_DAYS = 120


async def _assert_no_bookings(session: AsyncSession, column, value, what: str) -> None:
    existing = await session.execute(select(Booking.id).where(column == value).limit(1))
    if existing.scalar_one_or_none():
        raise HTTPException(409, f"Cannot delete a {what} with existing bookings")

router = APIRouter(prefix="/api/admin", tags=["admin-properties"])


@router.get("/properties")
async def list_properties_admin(
    session: AsyncSession = Depends(get_session),
    admin: AdminClaims = Depends(require_super_admin),
):
    result = await session.execute(select(Property))
    return result.scalars().all()


@router.post("/properties", response_model=Property)
async def create_property(
    payload: PropertyCreate,
    session: AsyncSession = Depends(get_session),
    admin: AdminClaims = Depends(require_super_admin),
):
    existing = await session.execute(select(Property).where(Property.slug == payload.slug))
    if existing.scalar_one_or_none():
        raise HTTPException(409, "A property with this slug already exists")
    property_ = Property(**payload.model_dump())
    session.add(property_)
    await session.commit()
    await session.refresh(property_)
    return property_


@router.patch("/properties/{property_id}", response_model=Property)
async def update_property(
    property_id: int,
    payload: PropertyUpdate,
    session: AsyncSession = Depends(get_session),
    admin: AdminClaims = Depends(require_super_admin),
):
    property_ = await session.get(Property, property_id)
    if not property_:
        raise HTTPException(404, "Property not found")
    updates = payload.model_dump(exclude_unset=True)
    if "slug" in updates and updates["slug"] != property_.slug:
        existing = await session.execute(select(Property).where(Property.slug == updates["slug"]))
        if existing.scalar_one_or_none():
            raise HTTPException(409, "A property with this slug already exists")
    for key, value in updates.items():
        setattr(property_, key, value)
    session.add(property_)
    await session.commit()
    await session.refresh(property_)
    return property_


@router.post("/room-types", response_model=RoomType)
async def create_room_type(
    payload: RoomTypeCreate,
    session: AsyncSession = Depends(get_session),
    admin: AdminClaims = Depends(require_super_admin),
):
    if not await session.get(Property, payload.property_id):
        raise HTTPException(404, "Property not found")
    room_type = RoomType(**payload.model_dump())
    session.add(room_type)
    await session.commit()
    await session.refresh(room_type)
    return room_type


@router.patch("/room-types/{room_type_id}", response_model=RoomType)
async def update_room_type(
    room_type_id: int,
    payload: RoomTypeUpdate,
    session: AsyncSession = Depends(get_session),
    admin: AdminClaims = Depends(require_super_admin),
):
    room_type = await session.get(RoomType, room_type_id)
    if not room_type:
        raise HTTPException(404, "Room type not found")
    for key, value in payload.model_dump(exclude_unset=True).items():
        setattr(room_type, key, value)
    session.add(room_type)
    await session.commit()
    await session.refresh(room_type)
    return room_type


@router.post("/rate-plans", response_model=RatePlan)
async def create_rate_plan(
    payload: RatePlanCreate,
    session: AsyncSession = Depends(get_session),
    admin: AdminClaims = Depends(require_super_admin),
):
    if not await session.get(RoomType, payload.room_type_id):
        raise HTTPException(404, "Room type not found")
    rate_plan = RatePlan(**payload.model_dump())
    session.add(rate_plan)
    await session.commit()
    await session.refresh(rate_plan)
    return rate_plan


@router.patch("/rate-plans/{rate_plan_id}", response_model=RatePlan)
async def update_rate_plan(
    rate_plan_id: int,
    payload: RatePlanUpdate,
    session: AsyncSession = Depends(get_session),
    admin: AdminClaims = Depends(require_super_admin),
):
    rate_plan = await session.get(RatePlan, rate_plan_id)
    if not rate_plan:
        raise HTTPException(404, "Rate plan not found")
    updates = payload.model_dump(exclude_unset=True)
    valid_from = updates.get("valid_from", rate_plan.valid_from)
    valid_to = updates.get("valid_to", rate_plan.valid_to)
    if valid_to <= valid_from:
        raise HTTPException(400, "valid_to must be after valid_from")
    for key, value in updates.items():
        setattr(rate_plan, key, value)
    session.add(rate_plan)
    await session.commit()
    await session.refresh(rate_plan)
    return rate_plan


@router.delete("/rate-plans/{rate_plan_id}", status_code=204)
async def delete_rate_plan(
    rate_plan_id: int,
    session: AsyncSession = Depends(get_session),
    admin: AdminClaims = Depends(require_super_admin),
):
    rate_plan = await session.get(RatePlan, rate_plan_id)
    if not rate_plan:
        raise HTTPException(404, "Rate plan not found")
    await _assert_no_bookings(session, Booking.rate_plan_id, rate_plan_id, "rate plan")
    await session.delete(rate_plan)
    await session.commit()
    return Response(status_code=204)


@router.delete("/room-types/{room_type_id}", status_code=204)
async def delete_room_type(
    room_type_id: int,
    session: AsyncSession = Depends(get_session),
    admin: AdminClaims = Depends(require_super_admin),
):
    room_type = await session.get(RoomType, room_type_id)
    if not room_type:
        raise HTTPException(404, "Room type not found")
    await _assert_no_bookings(session, Booking.room_type_id, room_type_id, "room type")
    await session.execute(delete(RatePlan).where(RatePlan.room_type_id == room_type_id))
    await session.execute(delete(Availability).where(Availability.room_type_id == room_type_id))
    await session.delete(room_type)
    await session.commit()
    return Response(status_code=204)


@router.delete("/properties/{property_id}", status_code=204)
async def delete_property(
    property_id: int,
    session: AsyncSession = Depends(get_session),
    admin: AdminClaims = Depends(require_super_admin),
):
    property_ = await session.get(Property, property_id)
    if not property_:
        raise HTTPException(404, "Property not found")
    await _assert_no_bookings(session, Booking.property_id, property_id, "property")

    room_type_ids_result = await session.execute(
        select(RoomType.id).where(RoomType.property_id == property_id)
    )
    room_type_ids = room_type_ids_result.scalars().all()
    if room_type_ids:
        await session.execute(delete(RatePlan).where(RatePlan.room_type_id.in_(room_type_ids)))
        await session.execute(delete(Availability).where(Availability.room_type_id.in_(room_type_ids)))
    await session.execute(delete(RoomType).where(RoomType.property_id == property_id))
    await session.delete(property_)
    await session.commit()
    return Response(status_code=204)


@router.get("/room-types/{room_type_id}/availability")
async def get_room_type_availability(
    room_type_id: int,
    date_from: date,
    date_to: date,
    session: AsyncSession = Depends(get_session),
    admin: AdminClaims = Depends(require_super_admin),
):
    if not await session.get(RoomType, room_type_id):
        raise HTTPException(404, "Room type not found")
    if date_from >= date_to:
        raise HTTPException(400, "date_to must be after date_from")
    if (date_to - date_from).days > MAX_AVAILABILITY_RANGE_DAYS:
        raise HTTPException(400, f"Range cannot exceed {MAX_AVAILABILITY_RANGE_DAYS} days")

    days = [date_from + timedelta(days=i) for i in range((date_to - date_from).days)]
    result = await session.execute(
        select(Availability).where(
            Availability.room_type_id == room_type_id, Availability.date.in_(days)
        )
    )
    by_date = {row.date: row.rooms_available for row in result.scalars().all()}
    return [{"date": d, "rooms_available": by_date.get(d, 0)} for d in days]


@router.put("/room-types/{room_type_id}/availability")
async def upsert_room_type_availability(
    room_type_id: int,
    payload: AvailabilityBulkUpsert,
    session: AsyncSession = Depends(get_session),
    admin: AdminClaims = Depends(require_super_admin),
):
    if not await session.get(RoomType, room_type_id):
        raise HTTPException(404, "Room type not found")

    dates = [d.date for d in payload.days]
    existing_result = await session.execute(
        select(Availability).where(
            Availability.room_type_id == room_type_id, Availability.date.in_(dates)
        )
    )
    existing_by_date = {row.date: row for row in existing_result.scalars().all()}

    for day in payload.days:
        row = existing_by_date.get(day.date)
        if row:
            row.rooms_available = day.rooms_available
            session.add(row)
        else:
            session.add(
                Availability(room_type_id=room_type_id, date=day.date, rooms_available=day.rooms_available)
            )
    await session.commit()
    return {"updated": len(payload.days)}

File: backend/src/routes/admin/properties.ts
import { Router } from "express";
import { query, queryOne, withTransaction } from "../../db";
import { asyncHandler, HttpError } from "../../middleware/errors";
import { validateBody } from "../../middleware/validate";
import {
  AvailabilityBulkUpsertSchema,
  PropertyCreateSchema,
  PropertyUpdateSchema,
  RatePlanCreateSchema,
  RatePlanUpdateSchema,
  RoomTypeCreateSchema,
  RoomTypeUpdateSchema,
} from "../../schemas/adminContent";
import { requireSuperAdmin } from "../../security";
import { Availability, Property, RatePlan, RoomType } from "../../types/models";

export const router = Router();

const MAX_AVAILABILITY_RANGE_DAYS = 120;

function addDays(dateStr: string, days: number): string {
  const d = new Date(dateStr + "T00:00:00Z");
  d.setUTCDate(d.getUTCDate() + days);
  return d.toISOString().slice(0, 10);
}

function daysBetween(a: string, b: string): number {
  return Math.round((Date.parse(b + "T00:00:00Z") - Date.parse(a + "T00:00:00Z")) / 86_400_000);
}

async function assertNoBookings(column: "property_id" | "room_type_id" | "rate_plan_id", value: number, what: string): Promise<void> {
  const existing = await queryOne(`SELECT id FROM bookings WHERE ${column} = $1 LIMIT 1`, [value]);
  if (existing) throw new HttpError(409, `Cannot delete a ${what} with existing bookings`);
}

function cleanAmenities(amenities: string[]): string[] {
  const cleaned = amenities.map((a) => a.trim().slice(0, 60)).filter((a) => a.length > 0);
  if (cleaned.length !== amenities.length) throw new HttpError(422, "Amenity labels cannot be blank");
  return cleaned;
}

router.use(requireSuperAdmin);

router.get(
  "/properties",
  asyncHandler(async (_req, res) => {
    const rows = await query<Property>("SELECT * FROM properties ORDER BY id");
    res.json(rows);
  })
);

router.post(
  "/properties",
  validateBody(PropertyCreateSchema),
  asyncHandler(async (req, res) => {
    const p = req.body as {
      name: string;
      slug: string;
      city: string;
      state: string;
      description: string;
      cover_image_url: string;
      latitude: number | null;
      longitude: number | null;
      amenities: string[];
    };
    const existing = await queryOne("SELECT id FROM properties WHERE slug = $1", [p.slug]);
    if (existing) throw new HttpError(409, "A property with this slug already exists");

    const amenities = cleanAmenities(p.amenities);
    const created = await queryOne<Property>(
      `INSERT INTO properties (name, slug, city, state, description, cover_image_url, latitude, longitude, amenities)
       VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9) RETURNING *`,
      [p.name, p.slug, p.city, p.state, p.description, p.cover_image_url, p.latitude, p.longitude, JSON.stringify(amenities)]
    );
    res.json(created);
  })
);

router.patch(
  "/properties/:propertyId",
  validateBody(PropertyUpdateSchema),
  asyncHandler(async (req, res) => {
    const propertyId = parseInt(req.params.propertyId, 10);
    const existing = await queryOne<Property>("SELECT * FROM properties WHERE id = $1", [propertyId]);
    if (!existing) throw new HttpError(404, "Property not found");

    const updates = req.body as Partial<Property>;
    if (updates.slug !== undefined && updates.slug !== existing.slug) {
      const clash = await queryOne("SELECT id FROM properties WHERE slug = $1", [updates.slug]);
      if (clash) throw new HttpError(409, "A property with this slug already exists");
    }
    const amenities = updates.amenities !== undefined ? cleanAmenities(updates.amenities) : existing.amenities;
    const merged = { ...existing, ...updates, amenities };

    const updated = await queryOne<Property>(
      `UPDATE properties SET name=$1, slug=$2, city=$3, state=$4, description=$5, cover_image_url=$6,
       latitude=$7, longitude=$8, amenities=$9, updated_at=now() WHERE id=$10 RETURNING *`,
      [
        merged.name,
        merged.slug,
        merged.city,
        merged.state,
        merged.description,
        merged.cover_image_url,
        merged.latitude,
        merged.longitude,
        JSON.stringify(merged.amenities),
        propertyId,
      ]
    );
    res.json(updated);
  })
);

router.post(
  "/room-types",
  validateBody(RoomTypeCreateSchema),
  asyncHandler(async (req, res) => {
    const p = req.body as { property_id: number; name: string; description: string; max_occupancy: number; base_price: number; images: string[] };
    const property = await queryOne("SELECT id FROM properties WHERE id = $1", [p.property_id]);
    if (!property) throw new HttpError(404, "Property not found");
    const created = await queryOne<RoomType>(
      `INSERT INTO room_types (property_id, name, description, max_occupancy, base_price, images)
       VALUES ($1,$2,$3,$4,$5,$6) RETURNING *`,
      [p.property_id, p.name, p.description, p.max_occupancy, p.base_price, JSON.stringify(p.images)]
    );
    res.json(created);
  })
);

router.patch(
  "/room-types/:roomTypeId",
  validateBody(RoomTypeUpdateSchema),
  asyncHandler(async (req, res) => {
    const roomTypeId = parseInt(req.params.roomTypeId, 10);
    const existing = await queryOne<RoomType>("SELECT * FROM room_types WHERE id = $1", [roomTypeId]);
    if (!existing) throw new HttpError(404, "Room type not found");
    const updates = req.body as Partial<RoomType>;
    const merged = { ...existing, ...updates };
    const updated = await queryOne<RoomType>(
      `UPDATE room_types SET name=$1, description=$2, max_occupancy=$3, base_price=$4, images=$5, updated_at=now() WHERE id=$6 RETURNING *`,
      [merged.name, merged.description, merged.max_occupancy, merged.base_price, JSON.stringify(merged.images), roomTypeId]
    );
    res.json(updated);
  })
);

router.post(
  "/rate-plans",
  validateBody(RatePlanCreateSchema),
  asyncHandler(async (req, res) => {
    const p = req.body as {
      room_type_id: number;
      name: string;
      fixed_price: number;
      refundable: boolean;
      includes_breakfast: boolean;
      valid_from: string;
      valid_to: string;
    };
    const roomType = await queryOne("SELECT id FROM room_types WHERE id = $1", [p.room_type_id]);
    if (!roomType) throw new HttpError(404, "Room type not found");
    const created = await queryOne<RatePlan>(
      `INSERT INTO rate_plans (room_type_id, name, fixed_price, refundable, includes_breakfast, valid_from, valid_to)
       VALUES ($1,$2,$3,$4,$5,$6,$7) RETURNING *`,
      [p.room_type_id, p.name, p.fixed_price, p.refundable, p.includes_breakfast, p.valid_from, p.valid_to]
    );
    res.json(created);
  })
);

router.patch(
  "/rate-plans/:ratePlanId",
  validateBody(RatePlanUpdateSchema),
  asyncHandler(async (req, res) => {
    const ratePlanId = parseInt(req.params.ratePlanId, 10);
    const existing = await queryOne<RatePlan>("SELECT * FROM rate_plans WHERE id = $1", [ratePlanId]);
    if (!existing) throw new HttpError(404, "Rate plan not found");
    const updates = req.body as Partial<RatePlan>;
    const validFrom = updates.valid_from ?? existing.valid_from;
    const validTo = updates.valid_to ?? existing.valid_to;
    if (validTo <= validFrom) throw new HttpError(400, "valid_to must be after valid_from");
    const merged = { ...existing, ...updates };
    const updated = await queryOne<RatePlan>(
      `UPDATE rate_plans SET name=$1, fixed_price=$2, refundable=$3, includes_breakfast=$4, valid_from=$5, valid_to=$6 WHERE id=$7 RETURNING *`,
      [merged.name, merged.fixed_price, merged.refundable, merged.includes_breakfast, merged.valid_from, merged.valid_to, ratePlanId]
    );
    res.json(updated);
  })
);

router.delete(
  "/rate-plans/:ratePlanId",
  asyncHandler(async (req, res) => {
    const ratePlanId = parseInt(req.params.ratePlanId, 10);
    const existing = await queryOne<RatePlan>("SELECT * FROM rate_plans WHERE id = $1", [ratePlanId]);
    if (!existing) throw new HttpError(404, "Rate plan not found");
    await assertNoBookings("rate_plan_id", ratePlanId, "rate plan");
    await query("DELETE FROM rate_plans WHERE id = $1", [ratePlanId]);
    res.status(204).send();
  })
);

router.delete(
  "/room-types/:roomTypeId",
  asyncHandler(async (req, res) => {
    const roomTypeId = parseInt(req.params.roomTypeId, 10);
    const existing = await queryOne<RoomType>("SELECT * FROM room_types WHERE id = $1", [roomTypeId]);
    if (!existing) throw new HttpError(404, "Room type not found");
    await assertNoBookings("room_type_id", roomTypeId, "room type");
    await withTransaction(async (client) => {
      await client.query("DELETE FROM rate_plans WHERE room_type_id = $1", [roomTypeId]);
      await client.query("DELETE FROM availability WHERE room_type_id = $1", [roomTypeId]);
      await client.query("DELETE FROM room_types WHERE id = $1", [roomTypeId]);
    });
    res.status(204).send();
  })
);

router.delete(
  "/properties/:propertyId",
  asyncHandler(async (req, res) => {
    const propertyId = parseInt(req.params.propertyId, 10);
    const existing = await queryOne<Property>("SELECT * FROM properties WHERE id = $1", [propertyId]);
    if (!existing) throw new HttpError(404, "Property not found");
    await assertNoBookings("property_id", propertyId, "property");

    await withTransaction(async (client) => {
      const roomTypeIdsRes = await client.query<{ id: number }>("SELECT id FROM room_types WHERE property_id = $1", [propertyId]);
      const roomTypeIds = roomTypeIdsRes.rows.map((r) => r.id);
      if (roomTypeIds.length) {
        await client.query("DELETE FROM rate_plans WHERE room_type_id = ANY($1::int[])", [roomTypeIds]);
        await client.query("DELETE FROM availability WHERE room_type_id = ANY($1::int[])", [roomTypeIds]);
      }
      await client.query("DELETE FROM room_types WHERE property_id = $1", [propertyId]);
      await client.query("DELETE FROM properties WHERE id = $1", [propertyId]);
    });
    res.status(204).send();
  })
);

router.get(
  "/room-types/:roomTypeId/availability",
  asyncHandler(async (req, res) => {
    const roomTypeId = parseInt(req.params.roomTypeId, 10);
    const dateFrom = String(req.query.date_from ?? "");
    const dateTo = String(req.query.date_to ?? "");
    if (!(await queryOne("SELECT id FROM room_types WHERE id = $1", [roomTypeId]))) throw new HttpError(404, "Room type not found");
    if (dateFrom >= dateTo) throw new HttpError(400, "date_to must be after date_from");
    if (daysBetween(dateFrom, dateTo) > MAX_AVAILABILITY_RANGE_DAYS) {
      throw new HttpError(400, `Range cannot exceed ${MAX_AVAILABILITY_RANGE_DAYS} days`);
    }

    const days: string[] = [];
    for (let i = 0; i < daysBetween(dateFrom, dateTo); i++) days.push(addDays(dateFrom, i));

    const rows = await query<Availability>("SELECT * FROM availability WHERE room_type_id = $1 AND date = ANY($2::date[])", [
      roomTypeId,
      days,
    ]);
    const byDate = new Map(rows.map((r) => [r.date, r.rooms_available]));
    res.json(days.map((d) => ({ date: d, rooms_available: byDate.get(d) ?? 0 })));
  })
);

router.put(
  "/room-types/:roomTypeId/availability",
  validateBody(AvailabilityBulkUpsertSchema),
  asyncHandler(async (req, res) => {
    const roomTypeId = parseInt(req.params.roomTypeId, 10);
    if (!(await queryOne("SELECT id FROM room_types WHERE id = $1", [roomTypeId]))) throw new HttpError(404, "Room type not found");

    const days = req.body.days as { date: string; rooms_available: number }[];
    await withTransaction(async (client) => {
      for (const day of days) {
        await client.query(
          `INSERT INTO availability (room_type_id, date, rooms_available) VALUES ($1,$2,$3)
           ON CONFLICT (room_type_id, date) DO UPDATE SET rooms_available = EXCLUDED.rooms_available`,
          [roomTypeId, day.date, day.rooms_available]
        );
      }
    });
    res.json({ updated: days.length });
  })
);
